Add mechanism to signal secure vs. insecure MQTT

andrewbonney opened the pull request:
This is one option to resolve #80 if we choose to.

I haven't included authorization as yet, but this brings MQTT into line with what is visible for WebSockets.

I've used 'broker_scheme' for now but I'm not wedded to that. 'broker_protocol' would have been more consistent with what we use elsewhere, but 'protocol' was never an ideal choice there.

andrewbonney committed:
Add mechanism to signal secure vs. insecure MQTT

garethsb-sony commented:
Yeah, I think I'd have gone for `broker_protocol` for consistency. 🤷‍♂ ([IANA call it a 'service name'](https://www.iana.org/assignments/service-names-port-numbers/service-names-port-numbers.xhtml?search=mqtt) for extra confusion.)

I suppose we have to have an enum? Means a version-up to add "ws" and "wss" support.

Also makes me wonder again why we don't call them `broker_host` and `broker_port` on both Senders and Receivers, rather than having `destination_` and `source_` variants.

andrewbonney commented:
> Yeah, I think I'd have gone for `broker_protocol` for consistency.  ([IANA call it a 'service name'](https://www.iana.org/assignments/service-names-port-numbers/service-names-port-numbers.xhtml?search=mqtt) for extra confusion.)

I'll change that

> I suppose we have to have an enum? Means a version-up to add "ws" and "wss" support.

Is this likely to be necessary? I don't know the details of this, but wouldn't MQTT over WebSocket use the WebSocket IS-05 type? If so that already has the means to specify a scheme.

> Also makes me wonder again why we don't call them `broker_host` and `broker_port` on both Senders and Receivers, rather than having `destination_` and `source_` variants.

This was purely for consistency with the already defined RTP parameters.

andrewbonney committed:
Modify MQTT param to use protocol rather than scheme

garethsb-sony commented:
> > I suppose we have to have an enum? Means a version-up to add "ws" and "wss" support.
> 
> Is this likely to be necessary? I don't know the details of this, but wouldn't MQTT over WebSocket use the WebSocket IS-05 type? If so that already has the means to specify a scheme.

Yes, I've been reading MQTT specs but not thinking hard enough - I think it's unnecessary, because MQTT over WebSocket is only likely to be used in browser-based MQTT clients (see e.g. https://www.hivemq.com/blog/mqtt-essentials-special-mqtt-over-websockets/).

FWIW, the WebSocket IS-05 type wouldn't make sense, however. That's point-to-point between Sender (server) and Receiver (client), whereas MQTT over WebSocket is still two connections: Receiver (client) to broker (server) _and_ Sender (client) to broker (server).

garethsb-sony commented:
One more comment: "plain text" to describe non-TLS is a bit confusing because MQTT is a binary protocol.

andrewbonney committed:
schemas: resolve review comment

andrewbonney commented:
I need to add support for 'auto' to these parameters too in case they're using DNS-SD.

andrewbonney committed:
mqtt: permit auto for broker_protocol parameters

andrewbonney committed:
Add missing 'broker_protocol' to constraints and examples

add missing

andrewbonney committed:
Add proposed parameters for authorization for MQTT/WebSocket

andrewbonney commented:
Fixed an omissions from the previous commits, and added a proposal for authorization parameters (needs discussion). Further consideration of use of 'auto' particularly for authorization parameters with WebSockets would be a good idea.

garethsb commented on APIs/schemas/receiver_transport_params_mqtt.json:
```diff
@@ -53,6 +53,18 @@
+              "string",
+              "boolean"
+            ],
+            "description": "Indication of whether authorization is used for communication with the broker. If the parameter is set to auto the Receiver should establish for itself whether authorization should be used, based on a discovery mechanism or its own internal configuration. ",
```
Pedant mode: there's an extra space on the end of the description string!

garethsb replied:
Some people may like `"auto"` for this to mean follow-this-API-authorization-setting (and similar logic when `"auto"` is specified for `"broker_protocol"`) but I think that's covered by 'internal configuration'?

andrewbonney committed:
schemas: remove trailing space

andrewbonney committed:
schemas: remove trailing spaces

andrewbonney committed:
examples: update protocol and authorization examples

andrewbonney committed:
examples: fix errors and omissions

garethsb commented on APIs/schemas/receiver_transport_params_mqtt.json:
```diff
@@ -44,6 +44,27 @@
+            "enum": [
+              "auto",
+              true,
+              false,
```
Extra comma

garethsb commented on APIs/schemas/receiver_transport_params_websocket.json:
```diff
@@ -24,6 +24,18 @@
+            "enum": [
+              "auto",
+              true,
+              false,
```
Extra comma

garethsb commented on APIs/schemas/sender_transport_params_mqtt.json:
```diff
@@ -39,11 +39,32 @@
+            "enum": [
+              "auto",
+              true,
+              false,
```
Extra comma

garethsb commented on APIs/schemas/sender_transport_params_websocket.json:
```diff
@@ -27,6 +27,18 @@
+            "enum": [
+              "auto",
+              true,
+              false,
```
...

andrewbonney committed:
examples: fix unnecessary trailing commas

garethsb commented on examples/sender-constraints-get-200-websocket-ext.json:
```diff
@@ -4,6 +4,12 @@
       "ws://172.29.80.65:5476/websocketServer"
     ]
   },
+  "connection_authorization": {
```
There's no point in this one - the empty object has identical behaviour?

garethsb commented on examples/sender-constraints-get-200-mqtt.json:
```diff
@@ -12,6 +12,18 @@
       "fixedTopicName2"
     ]
   },
+  "broker_protocol": {
```
There's no point in either of these ones - the empty object has identical behaviour?

garethsb commented on examples/receiver-constraints-get-200-websocket-ext.json:
```diff
@@ -1,5 +1,11 @@
 [{
   "connection_uri": {},
+  "connection_authorization": {
```
There's no point in this one - the empty object has identical behaviour?

andrewbonney committed:
examples: avoid unnecessary specifying of both boolean values
